osm/smi: update dir structure and add test

SanyaKochhar opened the pull request:
- Updates to collect osm custom resource meshconfig per latest osm release
- Fixes collection of smi crds (rbac in cluster-role)
- Adds naming structure to filename: `meshName/resourceName_resourceType` and `smi/resourceName_resourceType>`
- Tested with kustomize `kubectl apply -k deployment`

johnsonshi commented on deployment/aks-periscope.yaml:
```diff
@@ -28,6 +28,12 @@ rules:
 - apiGroups: ["apiextensions.k8s.io"]
   resources: ["customresourcedefinitions"]
   verbs: ["get", "list", "watch"]
+- apiGroups: ["access.smi-spec.io", "specs.smi-spec.io", "split.smi-spec.io"]
+  resources: ["*"]
+  verbs: ["get", "list", "watch"]
+- apiGroups: ["config.openservicemesh.io"]
+  resources: ["meshconfigs"]
+  verbs: ["get", "list", "watch"]
```
@SanyaKochhar I seem to recall @arnaud-tincelin and @Tatsinnit mentioning that the `deployment/aks-periscope.yaml` file is being deprecated. I was previously told to make `aks-periscope-role ClusterRole` changes in https://github.com/Azure/aks-periscope/blob/master/deployment/cluster-role.yaml.

@Tatsinnit can you confirm?

SanyaKochhar replied:
I did add these rbac permissions to `cluster-role.yaml` originally and I agree, it seems to make most sense to have them there, but it seemed like the rbac permissions doesn't work then. i.e., the `kubectl get meshconfig -A` command returns 
> Failed to collect meshconfigs for mesh osm: run command in container: exit status 1

Seeing similar error messaging for the SMI collector currently. Not sure if I'm missing some additional connection? @arnaud-tincelin @Tatsinnit please advise

SanyaKochhar replied:
Discussed this with @Tatsinnit yesterday. I'll be moving the meshconfig ClusterRole to cluster-role.yaml so it can be deployed using Kustomize. Still testing this -- running into some issues with Kustomize currently

SanyaKochhar replied:
Fixed this, moved cluster roles to cluster-role.yaml and tested this change with kustomize

Tatsinnit commented on pkg/collector/osm_collector.go:
```diff
@@ -252,10 +253,17 @@ func (collector *OsmCollector) collectGroundTruth(meshName string) {
-	collector.data[meshName+"_all_resources_configs"] = allResourcesConfigs
-	collector.data[meshName+"_mutating_webhook_configurations"] = mutationWebhookConfig
-	collector.data[meshName+"_validating_webhook_configurations"] = validatingWebhookConfig
+	meshConfig, err := utils.RunCommandOnContainer("kubectl", "get", "meshconfigs", "--all-namespaces", "-o", "json")
```
📔  Please note: for future work on this, lets make sure latter after GA of your need, we should plan a move towards using the `Client-go` packages into this collector, that will add better test-ability as well. https://github.com/kubernetes/client-go

There are some recent refactors we have done for those reason, one of the sample implementation is in ARC Pod Container work or system-err collector.

SanyaKochhar replied:
Sure, that makes sense

SanyaKochhar replied:
Added a test that I will also update once we refactor for client-go. Currently it's hard to test without client go.

Tatsinnit reviewed (changes requested):
💡 Thank you so much for this PR and pining us for review.

I have added few key comments, I assume you have test all this for `SMI\OSM` scenario and all works as expected. 

* If the `yaml` changes are specific to `SMIOSM` needs you might only want to keep them in your consuming tools.
* Please add some unit-test for this collector if possible, currently it decreases the over-all code-coverage. ( ![image](https://user-images.githubusercontent.com/6233295/134282000-73336bda-4171-47e1-8896-8dd57c7c66ce.png) )
* Regarding the over all, after you are GA, we should align a work for converting it using `client-go` package.

Thank you

SanyaKochhar commented on pkg/collector/osm_collector.go:
```diff
@@ -197,8 +198,8 @@ func (collector *OsmCollector) collectDataFromEnvoys(namespace string) error {
 			secretRemovedResponse := re.ReplaceAllString(string(responseBody), "---redacted---")
-
-			collector.data[query+"_"+podName] = secretRemovedResponse
+			filePath := meshName + "/envoy/"+ podName + query
```
There can be multiple meshes in a cluster and the collector collects all resources for each of them, so the number of files was getting overwhelming and had to discern between. Passing in `mesh` allows the file system to be organized so that resources collected for each mesh are grouped together, at least. Initially I was trying to just use a smarter naming system without creating subdirectories but the names were getting very long and unwieldy, ex: `<meshName>_namespace_<namespaceName>_pod_<podName>_logs`. If your team is open to it, I would prefer to include further subdirectories for each mesh, namespace, etc. as we had before the code was refactored so we can avoid such complex naming

cc: @arnaud-tincelin

SanyaKochhar commented on pkg/collector/osm_collector.go:
```diff
@@ -220,8 +221,8 @@ func (collector *OsmCollector) collectPodLogs(namespace string) error {
 		}
-
-		collector.data[podName+"_logs"] = output
+		filePath := meshName + "/" + podName + "_podLogs"
```
yes, more context here: https://github.com/Azure/aks-periscope/pull/118#discussion_r714008429

SanyaKochhar commented on deployment/cluster-role.yaml:
```diff
@@ -12,6 +12,12 @@ rules:
 - apiGroups: ["apiextensions.k8s.io"]
   resources: ["customresourcedefinitions"]
   verbs: ["get", "list", "watch"]
+- apiGroups: ["admissionregistration.k8s.io"]
```
I can move these out, I'm just wondering -- removing this rbac will mean that the SMI and OSM collectors will not completely work out of the box if someone were to try to use them. Is that something we want? An alternative would be to add some documentation around the kustomize changes required to use those two collectors

Tatsinnit reviewed (changes requested):
💡 🙏 Thanks for this, there are few comments, especially the yaml changes are something which need re-look. I see mainly PR is just the name and folder structure change and if you have tested this and only thing left is the `yaml` changes which is not needed in the core.

Can you please open a tracking item for the conversion of these collector to `client-go` we have already implemented logic `clients` and `podDecribe` For example a sample reside here: https://github.com/Azure/aks-periscope/blob/master/pkg/collector/kubeobjects_collector.go#L54

for future purpose regarding the `client-go > api` based implementation, I can help and implement this the conversion with you but I will need someone who understand the SMI/OSM scenario to test them alongside. 🙏☕️

Thank you so much,

Tatsinnit commented on pkg/collector/osm_collector_test.go:
```diff
@@ -0,0 +1,50 @@
+		name    string
+		want    int
+		wantErr bool
+		deployments []*appsv1.Deployment
```
💡 Idea: if you add `collectorName` like in other tests you could test the `GetName()` as well.

so add `collectorName string` here then, below  `value` --> `collectorName "osm"` later under the the test just do this around `line 47`: 

```
        name := c.GetName()
	if name != tt.collectorName {
		t.Errorf("GetName()) = %v, want %v", name, tt.collectorName)
	}
```

SanyaKochhar commented on deployment/cluster-role.yaml:
```diff
@@ -12,6 +12,3 @@ rules:
 - apiGroups: ["apiextensions.k8s.io"]
   resources: ["customresourcedefinitions"]
   verbs: ["get", "list", "watch"]
-- apiGroups: ["access.smi-spec.io", "specs.smi-spec.io", "split.smi-spec.io"]
```
It's SMI specific, which is why I have removed it from here

SanyaKochhar commented:
> 💡 🙏 Thanks for this, there are few comments, especially the yaml changes are something which need re-look. I see mainly PR is just the name and folder structure change and if you have tested this and only thing left is the `yaml` changes which is not needed in the core. Thank you for adding very initial test class for this as well.
> 
> Can you please open a tracking item for the conversion of these collector to `client-go` we have already implemented logic `clients` and `podDecribe` For example a sample reside here: https://github.com/Azure/aks-periscope/blob/master/pkg/collector/kubeobjects_collector.go#L54
> 
> for future purpose regarding the `client-go > api` based implementation, I can help and implement this the conversion with you but I will need someone who understand the SMI/OSM scenario to test them alongside. 🙏☕️
> 
> Thank you so much,

That is correct, I've now updated the title of the PR to reflect the same and opened an issue to track the conversion to client go here #124. Thanks for your help with this.

Tatsinnit reviewed (approved):
👍
